internal/storage/fs/git/store.go

```diff
@@ -98,18 +89,18 @@ func WithCABundle(caCertBytes []byte) containers.Option[SnapshotStore] {
 // fs.FS implementations around a target git repository.
 func NewSnapshotStore(ctx context.Context, logger *zap.Logger, url string, opts ...containers.Option[SnapshotStore]) (_ *SnapshotStore, err error) {
 	store := &SnapshotStore{
-		Closer: io.NopCloser(nil),
-		logger: logger.With(zap.String("repository", url)),
-		url:    url,
-		ref:    "main",
+		logger:  logger.With(zap.String("repository", url)),
+		url:     url,
+		baseRef: "main",
 	}
 	containers.ApplyAll(store, opts...)
 
-	field := zap.Stringer("ref", plumbing.NewBranchReferenceName(store.ref))
-	if store.hash != plumbing.ZeroHash {
-		field = zap.Stringer("SHA", store.hash)
+	store.logger = store.logger.With(zap.String("ref", store.baseRef))
+
+	store.snaps, err = storagefs.NewSnapshotCache[plumbing.Hash](logger, 3)
+	if err != nil {
+		return nil, err
 	}
-	store.logger = store.logger.With(field)
 
 	store.repo, err = git.Clone(memory.NewStorage(), nil, &git.CloneOptions{
```

Comment:
Shallow clones make subsequent fetches potentially very slow. This is because they don’t fetch any of the previous blobs or commits, which then have to be recomputed in the face of a subsequent fetch where you have partial information.

https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/#

> The other major difference is how git fetch behaves in a shallow clone. When fetching new commits, the server must provide every tree and blob that is “new” to these commits, relative to the shallow commits. This computation can be more expensive than a typical fetch, partly because a well-maintained server can make use of [reachability bitmaps](https://github.blog/2015-09-22-counting-objects/). Depending on how others are contributing to your remote repository, a git fetch operation in a shallow clone might end up downloading an almost-full commit history

Reply:
Potentially a blobless clone could work, though I suspect we might need to drop down to git itself and sidestep go git.
But a blobless clone would potentially allow us to only fetch the blob we want (things identified by the Flipt index). Sadly that blog doesn’t go into details about how to request specific blobs.